Make it an Incremental Source Generator

Hamunii opened the pull request:
As of right now, BepInEx.AutoPlugin is a non-incremental source generator. By making it incremental, it will feel much more responsive and hopefully will lead to less confusion, as it will regenerate the classes in real time when needed.

I might have modified more than I should about the project, but it hasn't been touched in a while so I did what I wanted to do, especially considering that this is basically a complete rewrite anyways.

Note: this extends #5 

Also worth noting that I implemented this too which I mentioned about previously (Edit: now removed since this isn't actually needed when taking properties from msbuild (Edit: this removal has been reverted due to changing back to taking info from attributes))

```xml
<PropertyGroup>
  <BepInAutoPluginStripBuildMetadata>true</BepInAutoPluginStripBuildMetadata>
</PropertyGroup>
```

Hamunii committed:
strip version metadata for BepInEx 5

Hamunii committed:
fix not setting version if not BepInEx 5

Hamunii committed:
make it incremental

Hamunii committed:
add .g.cs to BepInAutoPluginAttribute source

Hamunii committed:
target oldest roslyn version possible for maximum support

Hamunii committed:
remove probably stupid code

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
-";
-
-        private class SyntaxReceiver : ISyntaxContextReceiver
+    const string BepInAutoPluginAttr = "BepInEx.BepInAutoPluginAttribute";
+    const string PatcherAutoPluginAttr =
+        "BepInEx.Preloader.Core.Patching.PatcherAutoPluginAttribute";
+    const string BepInPluginAttr = "BepInEx.BepInPlugin";
+    const string PatcherPluginInfoAttr = "BepInEx.Preloader.Core.Patching.PatcherPluginInfo";
```
Do not use abbreviations in source code.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+            {
+                public PatcherAutoPluginAttribute(string? id = null, string? name = null, string? version = null) {}
+            }
         }
```
For robustness, prefix all namespaces with `global::`.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+                        out var stripMetadata
+                    );
+
+                    // Afaik these should always be set.
```
Not true. `<GenerateAssemblyInfo>false</GenerateAssemblyInfo>`

Hamunii replied:
Even if I set that to false, these values are still set to defaults which I think is the csproj file name and then 1.0.0 for the version. These values are exposed using the `<CompilerVisibleProperty Include="NameOfProperty" />` things in the `BepInEx.AutoPlugin.props` file, and those values are written to a `<NameOfProject>.GeneratedMSBuildEditorConfig.editorconfig` file where all the available properties are for us to use.

If the user however does not reference the `build` files from this package, then these values would be null.

Hamunii replied:
Okay reading again what I said, what I meant was that even with `<GenerateAssemblyInfo>false</GenerateAssemblyInfo>` set the info is still all there, but if I *don't set the relevant properties* in the csproj, then it will be the defaults which the SDK sets for those properties.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+                    );
+
+                    // Afaik these should always be set.
+                    assName ??= "unknown";
```
Like above, don't use abbreviations.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+        IncrementalValuesProvider<PluginClass> classesWithBepInAutoPlugin = context
+            .SyntaxProvider.ForAttributeWithMetadataName(
+                BepInAutoPluginAttr,
+                static (s, _) => s is ClassDeclarationSyntax,
```
```suggestion
                static (s, _) => s is ClassDeclarationSyntax { Parent: not TypeDeclarationSyntax } c && c.Modifiers.Any(SyntaxKind.PartialKeyword),
```

Hamunii replied:
I don't think this is a good idea since if you add the attribute to a class you are most likely intending to use this generator on that class, and in such a case we want compilation errors to tell the user that they are doing things wrong. There's also an analyzer providing a fix to make the plugin partial in such case

ds5678 replied:
A source generator generating bad code is always a bug. If you want there to be an error when the user's class isn't partial, you should have an analyzer report a diagnostic.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
-        public void Execute(GeneratorExecutionContext context)
+        var attr = ctx.Attributes.FirstOrDefault(att =>
+            att.AttributeClass?.ToDisplayString() == autoAttributeFqn
+        );
```
```suggestion
        var attribute = ctx.Attributes.First();
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+    public sealed record PluginClass(
+        string? Namespace,
+        string ClassName,
+        string BepInAttribute,
```
It is unnecessary to store this. Just use the constant directly.

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+                        /// <summary>
+                        /// The Id of <see cref="{{pluginClass}}"/>.
+                        /// </summary>
+                        public const string Id = "{{id}}";
```
Use `SymbolDisplay.FormatLiteral` for formatting C# strings.

ds5678 commented on BepInEx.AutoPlugin/BepInEx.AutoPlugin.csproj:
```diff
@@ -18,13 +15,28 @@
     <ItemGroup>
-        <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.9.0" PrivateAssets="all" />
-        <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.2" PrivateAssets="all" />
+        <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.3.1" PrivateAssets="all" />
+        <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3" PrivateAssets="all" />
```
Use the latest. There's no need to support outdated IDEs.

Hamunii replied:
Is there an advantage though to using references that are newer than necessary? If not, I think it makes sense to target the oldest that is new enough. Originally though I made this change (reverting my previous version upgrade) as there was a commit in this repo to reference older versions of these just to support older SDKs.

ds5678 replied:
Two reasons:
* Maintenance is easier when using latest. You never have to check if an API has changed.
* Enforcing a current version of the compiler ensures consistent behavior.

js6pak replied:
> Use the latest. There's no need to support outdated IDEs.

It's not only about IDEs, but also .NET sdks. 8.0.1xx is still supported and that ships with roslyn4.8

ds5678 replied:
Visual studio ships with .NET 9. Whenever I build anything, it uses that SDK. It only uses the reference assemblies from 6 and 8, unless a global.json explicitly dictates otherwise.

js6pak replied:
And everything except VS defaults to whatever SDK you have installed locally, which could be 8.0.1xx.

> unless a global.json explicitly dictates otherwise

Which is perfectly fine to do and why we should support the sdks which aren't EOL yet.

ds5678 replied:
So 14.8

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
+    static PluginClass? ToPluginClassWithAttribute(
+        GeneratorAttributeSyntaxContext ctx,
+        string attributeFqn,
+        string autoAttributeFqn
```
```suggestion
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,230 @@
         }
     }
+
+    public sealed record PluginClass(
```
Use a record struct instead.

Hamunii committed:
resolve most issues

Hamunii committed:
upgrade to latest CodeAnalysis references

Hamunii committed:
don't generate for invalid classes

Hamunii committed:
introduce analyzer for missing partial modifier

Hamunii commented:
I implemented an analyzer for the missing partial modifiers now. Not sure about its quality though since I have barely any experience with analyzers.

ds5678 commented on BepInEx.AutoPlugin.sln:
```diff
@@ -1,7 +1,10 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
+# 
```
Did you do this, or is it just automated formatting from the IDE?

Hamunii replied:
Automated formatting when I added the test plugin project

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
         {
-            Plugin,
-            Patcher,
+            return null;
```
```suggestion
            return default;
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
+        return true;
+    }
+
+    static PluginClass? ToPluginClass(GeneratorAttributeSyntaxContext ctx)
```
```suggestion
    static PluginClass ToPluginClass(GeneratorAttributeSyntaxContext ctx)
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
+                static (ctx, _) => ToPluginClass(ctx)
+            )
+            .Where(x => x is not null)
+            .Select((x, _) => (PluginClass)x!);
```
```suggestion
            .Where(x => x != default);
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
+                static (ctx, _) => ToPluginClass(ctx)
+            )
+            .Where(x => x is not null)
+            .Select((x, _) => (PluginClass)x!);
```
```suggestion
            .Where(x => x != default);
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
+
+    public record struct PluginClass(string? Namespace, string ClassName, PluginProps Overrides);
+
+    public record struct PluginProps(string? Id, string? Name, string? Version);
```
```suggestion
    public readonly record struct PluginClass(string? Namespace, string ClassName, PluginProps Overrides);

    public readonly record struct PluginProps(string? Id, string? Name, string? Version);
```

ds5678 commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,243 @@
+                    partial class {{pluginClass}}
                     {
-                        continue;
+                        /// <summary>
+                        /// The Id of <see cref="{{pluginClass}}"/>.
+                        /// </summary>
+                        public const string Id = {{id}};
+
+                        /// <summary>
+                        /// Gets the name of <see cref="{{pluginClass}}"/>.
+                        /// </summary>
+                        public static string Name => {{name}};
+
+                        /// <summary>
+                        /// Gets the version of <see cref="{{pluginClass}}"/>.
+                        /// </summary>
+                        public static string Version => {{version}};
```
Do you know why the first one is a constant, but the other two are static properties? It doesn't really make sense to me, other than maybe some use of reflection in BepinEx or an initial mistake which got baked into the output through a hesitance for breaking changes.

Hamunii replied:
The ID should never change, but the name and version don't have those restrictions and the version is expected to change, so users of an API using this can reference the properties and get an up to date value at runtime instead of at compile time

Hamunii committed:
more proper nullable handling

ds5678 commented on BepInEx.AutoPlugin/Analyzers/PluginClassAnalyzer.cs:
```diff
@@ -0,0 +1,83 @@
+    {
+        var classSyntax = (ClassDeclarationSyntax)context.Node;
+
+        var semanticModel = context.SemanticModel;
+        var compilation = context.Compilation;
+
+        INamedTypeSymbol?[] autoAttributes =
+        [
+            compilation.GetTypeByMetadataName(AutoPluginGenerator.BepInAutoPluginAttribute),
+            compilation.GetTypeByMetadataName(AutoPluginGenerator.PatcherAutoPluginAttribute),
+        ];
+
+        var attributes = classSyntax.AttributeLists.SelectMany(attrList => attrList.Attributes);
+        var hasAutoPluginAttribute = attributes.Any(attribute =>
+        {
+            if (semanticModel.GetSymbolInfo(attribute).Symbol is not IMethodSymbol attributeSymbol)
+                return false;
+
+            foreach (var autoAttribute in autoAttributes)
+            {
+                if (
+                    SymbolEqualityComparer.Default.Equals(
+                        attributeSymbol.ContainingSymbol,
+                        autoAttribute
+                    )
+                )
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        });
+
+        if (!hasAutoPluginAttribute)
+            return;
```
I think this can be done more efficiently. I have an example here:

https://github.com/ds5678/ds5678.Modding.TLD.SourceGenerator/blob/bc2a2d488a0c35aab9f1484c6888822e043a4e3d/ds5678.Modding.TLD.SourceGenerator/MarkerAttributeAnalyzer.cs#L26-L56

However, giving my code a second look, I see some areas it too could be improved.

Hamunii replied:
I implemented it very similar to your implementation now

ds5678 commented on BepInEx.AutoPlugin/Analyzers/PluginClassAnalyzer.cs:
```diff
@@ -0,0 +1,83 @@
+        if (!hasAutoPluginAttribute)
+            return;
+
+        if (classSyntax.Modifiers.Any(SyntaxKind.PartialKeyword))
+            return;
```
This is fast, so it should be before attribute checking.

ds5678 commented on BepInEx.AutoPlugin/Analyzers/PluginClassAnalyzer.cs:
```diff
@@ -0,0 +1,83 @@
+        if (classSyntax.Modifiers.Any(SyntaxKind.PartialKeyword))
+            return;
+
+        var diagnostic = Diagnostic.Create(
+            PluginClassMustBeMarkedPartial,
+            classSyntax.Identifier.GetLocation(),
+            classSyntax.Identifier.ToString()
+        );
```
This is good.

Hamunii committed:
change ToPluginClass nullability handling

Hamunii committed:
make record structs readonly

Hamunii committed:
downgrade Microsoft.CodeAnalysis.CSharp to oldest supported version

Hamunii committed:
this needs to be changed due to the downgraded version, whoops

Hamunii committed:
use ds5678's method for analyzer

Hamunii committed:
cleanup

- remove BepInAutoPluginStripBuildMetadata option since it's not actually needed when taking properties from msbuild
- made a method to get rid of essentially duplicate code for getting plugin classes
- moved comment about BepInEx 5 specific quirks to a more logical place

Hamunii committed:
update readme to be more comprehensive

Hamunii commented:
I would say that now this is in a pretty good shape as I think all issues are resolved (at least to me) and the readme is no longer outdated.

js6pak commented on BepInEx.AutoPlugin/AutoPluginGenerator.cs:
```diff
@@ -1,163 +1,236 @@
+                    var (options, isBepInEx5) = info;
+                    var globalOptions = options.GlobalOptions;
+                    globalOptions.TryGetValue("build_property.AssemblyName", out var assemblyName);
+                    globalOptions.TryGetValue("build_property.Product", out var productName);
```
What's the rationale for switching from `Title` to `Product`?
Are these properties even documented somewhere? The only difference between these two I see is that `Title` shows up in windows's file properties.

Hamunii replied:
Oh, yeah I didn't even realize this used `Title` and not `Product` like PluginInfoProps, so there is really no reason for the change. But now that I looked at it, this stackoverflow answer <https://stackoverflow.com/a/30219774> ~~makes it sound like `Product` would be the more appropriate property for the project name.~~ Actually never mind I saw an image of what it looks like, and I feel like `Title` would be very appropriate here.

Looking at how they behave, `Product` set is to a default value (seems like assembly name), but title is not set to anything.

Since PluginInfoProps uses `Product` and AutoPlugin has used `Title`, we could take `Title` first and check if it's defined, and if yes, use it, if not, use `Product` as the project name.

Hamunii committed:
use Title for project name with Product as a fallback

Hamunii committed:
take info from assembly attributes instead & reintroduce BepInAutoPluginStripBuildMetadata option

it turns out trying to take the info from MSBuild properties isn't reliable because why would it be.

Hamunii committed:
update readme to be up to date again

Hamunii commented:
Okay, so now the `Product` property is never used by AutoPlugin, and the logic is essentially the same as it used to be in the original version with using attributes, and so I reverted the removal of

```xml
<PropertyGroup>
  <BepInAutoPluginStripBuildMetadata>true</BepInAutoPluginStripBuildMetadata>
</PropertyGroup>
```

This uses InformationalVersionAttribute for the version even though I said it's wrong to use because of <https://learn.microsoft.com/en-us/dotnet/standard/assembly/versioning#assembly-informational-version> but honestly I don't care

Hamunii committed:
switch from using FAWMN for getting attribute arguments

This is probably slightly faster than using ForAttributeWithMetadataName based on my terrible benchmarks, and where FAWMN shines is attributes that don't exist, but these attributes should always exist.

Hamunii commented:
I switched off from using FAWMN since it seems like it's probably slightly faster this way.

Hamunii committed:
ignore all warnings in generated file

it's probably better to just ignore everything since the user can't do anything about it if they have some analyzer that that doesn't like what's happening there. One issue is when publicizing an assembly which also uses AutoPlugin.
